### DISCUSSION TITLE
FAQ example recommends an invalid use of Self

### DISCUSSION THREAD

**Discussion by dkamm** [beartype/beartype#247] [General]
I believe this FAQ example from "the current class" recommends an invalid use of `Self` because it returns a concrete type. See https://peps.python.org/pep-0673/#valid-locations-for-self

```python
# Import important stuff. Boilerplate: it's the stuff we make.
from beartype import beartype
from typing import Self  # <---------------- if Python ≥ 3.11.0
# from typing_extensions import Self   # <-- if Python < 3.11.0

# Decorate classes – not methods. It's rough.
@beartype  # <-- Yesss. Good. Feel the force. It flows like sweet honey.
class ClassFactory(object):
   def __init__(self, *args: Sequence) -> None:
       self._args = args

   # @beartype  # <-- No... Oh, Gods. *NO*! The dark side grows stronger.
   def make_class(self, other: Self) -> Self:  # <-- We are all one self.
       return ClassFactory(self._args + other._args)
```

Seems like Forward Reference or Postponed Evaluation is necessary for these cases

**Comment by leycec** [comment]
**Fascinating.** I was wondering when someone was going to notice how strongly @beartype diverges from strict conformance with PEP 673. To be blunt, **this is entirely intentional.** 

@beartype is a *pragmatic* type-checker; that's why it runs at runtime rather than static analysis time. @beartype intentionally rejects those aspects of typing standards that are actively harmful to real-world quality assurance. As example, @beartype rejects the [PEP 484-compliant implicit numeric tower](https://peps.python.org/pep-0484/#the-numeric-tower) by default, because accepting that by default would invite a silent loss of precision in numerical APIs, which would be actively harmful to real-world quality assurance; instead, @beartype allows end users to explicitly enable the implicit numeric tower by configuring @beartype accordingly (e.g., as `@beartype(conf=BeartypeConf(is_pep484_tower=True))`).

Static type-checkers like `mypy` and `pyright`, on the other hand, are *ideological* type-checkers; they enforce static typing on a dynamically typed language for purely ideological reasons that are actively harmful to real-world quality assurance. They dramatically increase the number of reported **false positives** (i.e., erroneous claims that a given module contains fatal typing errors when in fact that module is merely dynamically typed, because that module is Python). That only uselessly wastes everyone's scarce time.

## What Are You Talking About, @leycec?

*...heh.* I kinda just rambled incoherently there, didn't I? Please don't answer that.

What I am *trying* to say here is that most of the common use cases that PEP 673 rejects are actually valid use cases that PEP 673 should have instead accepted. These include:

* **Factory methods.** Type-checking that a factory method returns the expected object absolutely is one of the primary use cases for real-world `typing.Self` usage. @beartype therefore accepts rather than rejects this use case. Even if @beartype *wanted* to reject this use case (which we don't), we couldn't. Why? Because @beartype runs at runtime – not static analysis time. @beartype has *no* means of deterministically differentiating a factory method from a non-factory method. That's a semantic distinction that's entirely up to subjective human interpretation. @beartype doesn't particularly care what kind of method it's type-checking or how the object returned by that method was instantiated; it just unconditionally type-checks `Self` the same *everywhere*. Is the object being type-checked an instance of the current class? If so, great; it satisfies `Self`. If not, not great; it violates `Self`. That's it. Frankly, that's what every other type-checker should be doing, too. That way lies sanity. All other roads lead to madness.
* **Type variables** (e.g., `T = TypeVar('T')`). @beartype could mostly care less about type variables. Admittedly, that's awful. Ideally, @beartype *would* deeply care about type variables. But @beartype currently doesn't, because type variables were only ever intended to be cared about by static type-checkers. They're *really* hard to efficiently type-check at runtime in a way that doesn't destroy performance. Unsurprisingly, therefore, @beartype happily accepts this use case that PEP 673 tells us to reject:

```python
class Foo:
    # Rejected (Self is treated as unknown).
    def has_existing_self_annotation(self: T) -> Self: ...
```

I literally have *no* idea why @beartype should even reject that, in theory. What does `"Self is treated as unknown"` even *mean*? @beartype never treats `Self` as unknown. `Self` is *always* known, because the current class is always known. With respect to @beartype, `Self` just means "instance of the current class." That's it. Unsurprisingly, @beartype happily accepts the above.

* **Static methods.** PEP 673 tells us to reject static methods annotated as returning `Self`. Why? Again, I literally have no idea. PEP 673 waves its hands spasmodically and wants us to believe that "These don’t seem worth the additional complexity." What are you talking about, PEP 673? What "additional complexity"? I have no idea. In fact, rejecting static methods annotated as returning `Self` would require considerably *more* complexity than just silently accepting this as a valid use case. I rolled my eyes many times when I read that nonsensical pablum. So did @beartype.
* **Metaclasses.** Again, PEP 673 makes utterly *no* sense here. PEP 673 tells us to reject `Self` in metaclasses. Again, @beartype ignores PEP 673. Why? Because @beartype runs at runtime – not static analysis time. @beartype has *no* means of deterministically differentiating a metaclass from a class. Moreover, there's *no* demonstrable reason to attempt to do so. Metaclasses are literally just classes. PEP 673 tries really hard to make us believe that accepting `Self` in metaclasses would invite "confusion." No, there is *no* confusion, PEP 673. In any method of a metaclass, <sup>...except `__new__()`, of course, which is the exact same for standard classes</sup> `Self` refers to an instance of that metaclass (i.e., the current class whose type is that metaclass). If you understand metaclasses, you understand that this is no more confusing than metaclasses themselves.

## You... Kinda Have a Point There, @leycec

I know, right? I kinda do.

So... why didn't PEP 673 just accept those use cases then if they're supposedly so valid? **Because PEP 673 was authored only for static type-checkers.** The intended audience is `mypy` and `pyright` – *not* @beartype and `typeguard`. Type-checking that is hard at static analysis time is trivial at runtime and vice versa.

In the case of PEP 673, most of the use cases that PEP 673 rejects are rejected merely because they're hard for static type-checkers. But that's a bad reason to reject those use cases. The rest of the use cases that PEP 673 rejects are rejected purely for ideological reasons. But that's also a bad reason to reject those use cases.

Type-checking isn't about purity. It's about practicality. Purity is the enemy of type-checking, because purity invites nonsensical false positives that waste everyone's time. @beartype *always* errs on the side of practicality.

This is the way. Thus spake the bear. :muscle: :bear: 

Oh – and thanks so much for submitting this fascinating discussion topic, @dkamm. I *love* rambling at voluminous length about the intersection of philosophy and quality assurance. You can probably tell that I'm a utilitarian pragmatist at heart rather than a Platonic idealist. Go, go, greatest good for the greatest number! QA forevah.
